smart_card_connector_app/src/chrome-api-jstocxxtest.js

```diff
@@ -746,339 +591,481 @@ goog.exportSymbol('testChromeApiProviderToCpp', {
+        .$waitAndVerify();
+  },
+
+  // Test that Control fails for the `IOCTL_SMARTCARD_VENDOR_IFD_EXCHANGE`
+  // command.
+  'testControl_fail': async function() {
+    expectReportDataResult(
+        /*requestId=*/ 111, new ArrayBuffer(0), 'NOT_TRANSACTED');
+    await mockChromeApi
+        .dispatchEvent(
+            'onControlRequested', /*requestId=*/ 111, readerHandle,
+            IOCTL_SMARTCARD_VENDOR_IFD_EXCHANGE, new ArrayBuffer(0))
+        .$waitAndVerify();
+  },
+};
+
+goog.exportSymbol('testChromeApiProviderToCpp', {
+  'setUp': async function() {
+    // Set up the controller and load the C/C++ executable module.
+    testController = new GSC.IntegrationTestController();
+    await testController.initAsync();
+    // Stub out libusb receiver.
+    libusbProxyReceiver = new GSC.LibusbProxyReceiver(
+        testController.executableModule.getMessageChannel());
+    libusbProxyReceiver.addHook(new GSC.TestingLibusbSmartCardSimulationHook(
+        testController.executableModule.getMessageChannel()));
+    // Set up observers.
+    pcscReadinessTracker = new ReadinessTracker(
+        testController.executableModule.getMessageChannel());
+    // Mock chrome.smartCardProviderPrivate API.
+    mockControl = new goog.testing.MockControl();
+    mockChromeApi =
+        new MockChromeApi(mockControl, testController.propertyReplacer);
+    mockControl.$replayAll();
+
+    createChromeApiProvider();
+  },
+
+  'tearDown': async function() {
+    try {
+      await testController.disposeAsync();
+      pcscReadinessTracker.dispose();
+      assertTrue(chromeApiProvider.isDisposed());
+      // Check all mock expectations are satisfied.
+      mockControl.$verifyAll();
+    } finally {
+      chromeApiProvider = null;
+      pcscReadinessTracker = null;
+      testController = null;
+    }
+  },
+
+  ...SIMPLE_TEST_CASES,
 
+  // Tests that check that requests on an already released context return
+  // INVALID_HANDLE.
+  'testReleasedContext': {
+    'setUp': async function() {
+      await launchPcscServer(/*initialDevices=*/[]);
+      // Establish and release context.
+      releasedContext = await establishContext(/*requestId=*/ 123);
+      expectReportReleaseContext(/*requestId=*/ 124, 'SUCCESS');
       await mockChromeApi
           .dispatchEvent(
-              'onGetAttribRequested', /*requestId=*/ 111, readerHandle,
-              API.SCARD_ATTR_ATR_STRING)
+              'onReleaseContextRequested', /*requestId=*/ 124, releasedContext)
           .$waitAndVerify();
     },
 
-    // Test that SetAttrib fails for unsupported attributes.
-    'testSetAttrib_error': async function() {
-      expectReportPlainResult(/*requestId=*/ 111, 'INVALID_PARAMETER');
+    'tearDown': function() {
+      releasedContext = null;
+    },
 
+    // Test that ReleaseContext requested on an already released context
+    // returns INVALID_HANDLE error.
+    'testReleaseContext': async function() {
```

Comment:
Let's move these test cases outside, to separate variables as well? That'd add a structure to the whole test file: first test case functions, then setups+teardowns.

Reply:
done :)